consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -1,165 +1,463 @@
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("view-already-acknowledged", func() {
+		// create block with view <= HighestAcknowledgedView
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					func(block *model.Block) {
+						block.QC = helper.MakeQC(helper.WithQCView(s.safetyData.HighestAcknowledgedView - 1))
+					},
+					helper.WithBlockView(s.safetyData.HighestAcknowledgedView))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.True(s.T(), model.IsNoVoteError(err))
+		require.Nil(s.T(), vote)
+	})
+	s.Run("no-last-view-tc", func() {
+		// create block where Block.View != Block.QC.View+1 and LastViewTC = nil
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("last-view-tc-invalid-view", func() {
+		// create block where Block.View != Block.QC.View+1 and
+		// Block.View != LastViewTC.View+1
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))),
+			helper.WithLastViewTC(
+				helper.MakeTC(
+					helper.WithTCView(s.bootstrapBlock.View))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+	s.Run("last-view-tc-invalid-highest-qc", func() {
+		// create block where Block.View != Block.QC.View+1 and
+		// Block.View == LastViewTC.View+1 and Block.QC.View < LastViewTC.TOHighestQC.View
+		// in this case block is not safe to extend since proposal is built on top of QC, which is lower
+		// than QC presented in LastViewTC.
+		TOHighestQC := helper.MakeQC(helper.WithQCView(s.bootstrapBlock.View + 1))
+		proposal := helper.MakeProposal(
+			helper.WithBlock(
+				helper.MakeBlock(
+					helper.WithParentBlock(s.bootstrapBlock),
+					helper.WithBlockView(s.bootstrapBlock.View+2))),
+			helper.WithLastViewTC(
+				helper.MakeTC(
+					helper.WithTCView(s.bootstrapBlock.View+1),
+					helper.WithTCHighestQC(TOHighestQC))))
+		vote, err := s.safety.ProduceVote(proposal, proposal.Block.View)
+		require.Error(s.T(), err)
+		require.Nil(s.T(), vote)
+	})
+
+	s.signer.AssertNotCalled(s.T(), "CreateVote")
+	s.persister.AssertNotCalled(s.T(), "PutSafetyData")
+}
 
-	// create voter
-	block, _, voter := createVoter(blockView, lastVotedView, isBlockSafe, isCommitteeMember)
+// TestProduceTimeout_ShouldTimeout tests that we can produce timeout in cases where
+// last view was successful or not. Also tests last timeout caching.
+func (s *SafetyRulesTestSuite) TestProduceTimeout_ShouldTimeout() {
```

Comment:
Suggestion for additional test cases:
* we should be able to produce a vote for a view, then timeout for the same view
* we should be unable to timeout for a view, then vote for the same view